cut the validation <-> txmempool circular dependency 2/2

glozow opened the pull request:
Remove 2 circular dependencies: validation - txmempool and validation - policy/rbf - txmempool

Validation should depend on txmempool (e.g. `CChainstateManager` has a mempool and we often need to know what's in our mempool to validate transactions), but txmempool is a data structure that shouldn't really need to know about chain state.

- Changes `removeForReorg()` to be parameterized by a callable that returns true/false (i.e. whether the transaction should be removed due to being now immature or nonfinal) instead of a `CChainState`. The mempool really shouldn't need to know about coinbase maturity or lockpoints, it just needs to know which entries to remove.

JeremyRubin commented:
Hmm I'm mildly concept NACK on this refactoring; removeForReorg really feels like internal mempool functionality.

Another tactic to remove the module links would be to make removeForReorg generic to any T ChainState class. Then you don't need to have the specifics of the implementation of T known within mempool.

(Circular deps within `.cpp` are not *that* bad IMO)

fanquake commented:
```bash
validation.cpp:338:25: error: calling function 'TestLockPointValidity' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
        bool validLP =  TestLockPointValidity(m_chain, &lp);
                        ^
validation.cpp:339:41: error: calling function 'CoinsTip' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
        CCoinsViewMemPool view_mempool(&CoinsTip(), pool);
                                        ^
validation.cpp:340:14: error: calling function 'CheckFinalTx' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
        if (!CheckFinalTx(m_chain.Tip(), tx, flags)
             ^
validation.cpp:350:36: error: calling function 'CoinsTip' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
                const Coin &coin = CoinsTip().AccessCoin(txin.prevout);
                                   ^
validation.cpp:5113:5: error: calling function 'AssertLockHeldInternal<AnnotatedMixin<std::__1::recursive_mutex> >' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
    AssertLockHeld(::cs_main);
    ^
./sync.h:81:28: note: expanded from macro 'AssertLockHeld'
#define AssertLockHeld(cs) AssertLockHeldInternal(#cs, __FILE__, __LINE__, &cs)
                           ^
validation.cpp:5117:10: error: calling function 'check' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
    pool.check();
         ^
validation.cpp:5119:41: error: calling function 'CoinsTip' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
    CCoinsViewCache& active_coins_tip = CoinsTip();
                                        ^
7 errors generated.
```

glozow commented:
> removeForReorg really feels like internal mempool functionality. Another tactic to remove the module links would be to make removeForReorg generic to any T ChainState class. Then you don't need to have the specifics of the implementation of T known within mempool.

Yeah, I see this. Perhaps the function could instead be parametrized by tip, coins cache, etc?

glozow commented:
I've pushed a version that splits out the finality checking part of removeForReorg (parts that require validation functions CheckSequenceLocks and CheckFinalTx) into validation, but leaves most of it in txmempool. Perhaps a better separation?

JeremyRubin commented:
i don't think so :/

glozow commented:
@JeremyRubin `CChainState` is what it's currently parametrized by - it's what causes the dependency on validation.h

glozow commented:
Ok new approach. I added something similar to the `CConnman.ForEachNode` so we can pass in a lambda that captures what we need from chainstate. Validation isn't looking through `mapTx` anymore. The mempool gets a callable object and applies it to the entries internally.

JeremyRubin commented:
i meant template paramterized by a generic ChainStateProvider type, that way it's clean depedency wise.. will check out the new approach later.

michaelfolkson commented:
What's the latest on this PR? Am I right in thinking that an approach hasn't been settled on to avoid the circular dependency and that #22675 should be reviewed instead?

glozow commented:
Fixed the CI issue. Ready for review.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2288,7 +2288,7 @@ void PeerManagerImpl::ProcessOrphanTx(std::set<uint256>& orphan_work_set)
             break;
         }
     }
-    m_mempool.check(m_chainman.ActiveChainstate());
```
It'd be nice to remove these calls to mempool.check() from net_processing. It shouldn't be the client's responsibility to call into validation/mempool to check its invariants.

The diff is really easy:

```diff
iff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 80655c61e7..636b4d1102 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -2288,7 +2288,6 @@ void PeerManagerImpl::ProcessOrphanTx(std::set<uint256>& orphan_work_set)
             break;
         }
     }
-    m_mempool.check(m_chainman.ActiveChainstate());
 }
 
 bool PeerManagerImpl::PrepareBlockFilterRequest(CNode& peer,
@@ -3250,7 +3249,6 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         const TxValidationState& state = result.m_state;
 
         if (result.m_result_type == MempoolAcceptResult::ResultType::VALID) {
-            m_mempool.check(m_chainman.ActiveChainstate());
             // As this version of the transaction was acceptable, we can forget about any
             // requests for it.
             m_txrequest.ForgetTxHash(tx.GetHash());
diff --git a/src/validation.cpp b/src/validation.cpp
index 2a66d96f22..8f1f33782b 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1041,7 +1041,9 @@ static MempoolAcceptResult AcceptToMemoryPoolWithTime(const CChainParams& chainp
 MempoolAcceptResult AcceptToMemoryPool(CChainState& active_chainstate, CTxMemPool& pool, const CTransactionRef& tx,
                                        bool bypass_limits, bool test_accept)
 {
-    return AcceptToMemoryPoolWithTime(Params(), pool, active_chainstate, tx, GetTime(), bypass_limits, test_accept);
+    auto ret = AcceptToMemoryPoolWithTime(Params(), pool, active_chainstate, tx, GetTime(), bypass_limits, test_accept);
+    pool.check(active_chainstate);
+    return ret;
 }
 
 PackageMempoolAcceptResult ProcessNewPackage(CChainState& active_chainstate, CTxMemPool& pool,
```
branch here: https://github.com/jnewbery/bitcoin/tree/2021-09-no-mempool-check-in-net-processing

That's a _slight_ behaviour change since check() will get called a little more often, so it doesn't fit inside this PR, but by doing it first, this PR wouldn't need to touch net_processing. Thoughts?

glozow replied:
Didn't end up taking this change, though I tried to at first, because I agree that it _should_ be the mempool, or at the very least, chainstate manager's job to sanity check the mempool, not net processing.

The problem is that we call ATMP from the reorg code, and we shouldn't call `check()` then because the mempool can temporarily be in a state where not all inputs are available from the chainstate coins cache. This causes the `check()` assertions to fail.

jnewbery replied:
> The problem is that we call ATMP from the reorg code, and we shouldn't call check() then because the mempool can temporarily be in a state where not all inputs are available from the chainstate coins cache. This causes the check() assertions to fail.

Yes, I'd forgotten that ATMP could be called while the mempool is in an inconsistent state.

The fix is to only call `check()` after ATMP has been called from outside validation. I have a branch that implements a `ChainstateManager.ProcessTransaction()` interface method which does exactly that: https://github.com/jnewbery/bitcoin/tree/2021-09-process-transaction. I'll probably go ahead and PR that soon unless you think it should wait until after this PR.

jnewbery reviewed (commented):
Concept ACK!

I've left a few comments inline. Another couple of comments on the git history:

1.

Commit 3 (_[mempool] split CTxMemPool::check into contextual and internal parts_) claims to be no behaviour change, but the new internal `check()` method is not called by external `check(CChainState& active_chainstate)` method. That is restored in a later commit, but it should be included in commit 3 to make the commits hygenic (and the later commit _MOVEONLY: contextual mempool check to validation_ would be more move-only)

2.

It'd be nice if the _MOVEONLY: contextual mempool check to validation_ commit had fewer non-move changes so that using `git diff --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space` would obviously show that the code really is just moving. Are you able to rename variables inside check(), so that the changes are minimal in the move. Same for the _MOVEONLY: filtering nonfinal/immature transactions to validation_ commit.

glozow committed:
[refactor] put finality and maturity checking into a lambda

No behavior change.

glozow committed:
Break validation <-> txmempool circular dependency

No behavior change.

Parameterize removeForReorg using a CChain and callable that
encapsulates validation logic.  The mempool shouldn't need to know a
bunch of details about coinbase maturity and lock finality. Instead,
just pass in a callable function that says true/false. Breaks circular
dependency by removing txmempool's dependency on validation.

glozow committed:
[mempool] always assert coin spent

This is an extremely cheap function (just checks that the coin CTxOut
isn't null) that doesn't need to be gated on check_ratio.

glozow commented on src/txmempool.cpp:
```diff
@@ -563,39 +562,11 @@ void CTxMemPool::removeRecursive(const CTransaction &origTx, MemPoolRemovalReaso
-            || !CheckSequenceLocks(active_chainstate.m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
-            // Note if CheckSequenceLocks fails the LockPoints may still be invalid
-            // So it's critical that we remove the tx and not depend on the LockPoints.
-            txToRemove.insert(it);
```
Good catch, switched it to be updating a `bool should_remove` and returning at the end instead :)

glozow commented on src/txmempool.cpp:
```diff
@@ -671,19 +642,9 @@ void CTxMemPool::clear()
 {
     if (m_check_ratio == 0) return;
-
     if (GetRand(m_check_ratio) >= 1) return;
```
keeping `check_ratio` checking inside the mempool function now

glozow commented on src/txmempool.cpp:
```diff
@@ -563,39 +562,11 @@ void CTxMemPool::removeRecursive(const CTransaction &origTx, MemPoolRemovalReaso
 }
 
-void CTxMemPool::removeForReorg(CChainState& active_chainstate, int flags)
+void CTxMemPool::removeForReorg(const setEntries& txToRemove)
```
Agree about conslidating duplicate code, but I don't think it helps much with this PR - maybe for a followup?

glozow commented:
I've changed the circular dependency-breaking to be as minimal as possible. The `check()` function is now parametrized by a coins cache and spend height rather than chainstate. The `removeForReorg()` function now takes a callable object to apply to each mempool entry, removing the ones for which it returns true.

I've also written a bench for these 2 functions and taken the opportunity to speed `check()` up by eliminating the need to check mempool entries more than once. My bench results show a 35% speedup. Hopefully the performance improvement + circular dependency break is good enough to motivate these changes.

Bench at commit when the benchmark is first introduced:
![image](https://user-images.githubusercontent.com/25183001/135457316-3829cabf-10fe-4b18-bf37-f411e24b3011.png)


Bench at HEAD of this PR:
![image](https://user-images.githubusercontent.com/25183001/135457337-da5bd713-401e-4ef2-a2db-2de37120a910.png)

jnewbery commented on src/validation.cpp:
```diff
@@ -367,8 +367,42 @@ void CChainState::MaybeUpdateMempoolForReorg(
+        AssertLockHeld(::cs_main);
+        const CTransaction& tx = it->GetTx();
+        LockPoints lp = it->GetLockPoints();
+        bool validLP = TestLockPointValidity(m_chain, &lp);
```
`TestLockPointValidity()` can now be static since it's not called from outside this translation unit.

glozow replied:
Done

glozow commented:
Rebased, addressed @jnewbery's comment, and separated the benches because it's a bit misleading to group them.

glozow commented:
Split the PR up - first half is in #23157.

mjdietzx reviewed (commented):
Code review ACK b5de6cc10bfbc8174936c3e8ae9418dedfb60000

Very nicely done as usual. I also did some light testing - built locally, unit tests pass, also ran the functional tests I added in #22867 and those still pass ✅

theStack reviewed (approved):
Code-review ACK b5de6cc10bfbc8174936c3e8ae9418dedfb60000

TIL that lambda captures can also be used to define new variables/constant, rather than just capturing existing ones (as done for `flags`).

MarcoFalke commented:
Needs rebase?

glozow commented:
Thanks @MarcoFalke! silent merge conflict with #23211. Need to rethink approach since it's no longer possible to modify lockpoints from outside txmempool...

MarcoFalke commented:
That was a move-only, so can be reverted if needed. Maybe a `struct update_lock_points;` forward decl is enough?

glozow commented:
> That was a move-only, so can be reverted if needed. Maybe a `struct update_lock_points;` forward decl is enough?

Forward decl isn't enough because we need the constructor in order to call `update_lock_points(lp)`.

Currently, I'm leaning towards splitting `removeForReorg` into 2 loops: `CheckLockPoints` (leave inside txmempool) and `RemoveNonFinalAndPremature` (move to validation). There shouldn't be a performance penalty.

Reverting is also an option. I'm currently trying to figure out whether that's better. AFAICT, the `update_*` structs help provide an interface to modifying mempool entries while keeping the `txiter` type const (txmempool passes `txiters` and `setEntries`s to outside callers pretty freely). But since all modifications to entries should be done through txmempool's public member functions, perhaps it's better for them to be private. Will investigate further.

glozow committed:
[mempool] only update lockpoints for non-removed entries

Each entry's lockpoints are independent of one another, so there isn't
any reason to update lockpoints for entries that will be removed.
Separating the loops also allows us to move validation logic out and
leave lockpoints in txmempool.

glozow committed:
MOVEONLY: TestLockPointValidity to txmempool

glozow commented:
Rebased. If you have time, please take another look @mjdietzx @theStack @MarcoFalke?

laanwj commented:
Code review ACK a64078e38563ef3ac5e5ec20c07569441c87eeee

theStack commented on src/txmempool.h:
```diff
@@ -14,6 +14,7 @@
 #include <utility>
 #include <vector>
 
+#include <chain.h>
```
nit: could use the forward-declaration `class CChain;` instead to reduce header dependencies

glozow replied:
taken in #23649

theStack reviewed (approved):
re-ACK a64078e38563ef3ac5e5ec20c07569441c87eeee

mjdietzx commented:
reACK a64078e38563ef3ac5e5ec20c07569441c87eeee

MarcoFalke commented on src/txmempool.cpp:
```diff
@@ -663,15 +663,19 @@ void CTxMemPool::removeForReorg(CChainState& active_chainstate, int flags)
         CalculateDescendants(it, setAllRemoves);
     }
     RemoveStaged(setAllRemoves, false, MemPoolRemovalReason::REORG);
+    auto chain = active_chainstate.m_chain;
```
bedf246f1e2497a3641093c6e8fa11fb34dddac4:

Is this creating a copy of a vector holding 1kk elements for no reason?

glozow replied:
woop, should have declared as `auto&`. The good thing is it's removed in the last commit?

MarcoFalke commented on src/txmempool.cpp:
```diff
@@ -663,15 +663,19 @@ void CTxMemPool::removeForReorg(CChainState& active_chainstate, int flags)
+    auto chain = active_chainstate.m_chain;
+    for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
+        LockPoints lp = it->GetLockPoints();
+        if (!TestLockPointValidity(chain, &lp)) {
```
unrelated: Seems a bit confusing to pass a pointer here. Usually this means the memory it points to will be modified. However, in this case `lp` can be const.

MarcoFalke replied:
Actually bedf246f1e2497a3641093c6e8fa11fb34dddac4 is wrong.

`CheckSequenceLocks` modifies `lp`, so dropping it and using a fresh `lp` here from `GetLockPoints` is illegal.

glozow replied:
Oof, that's not good. Can be fixed by moving `update_lock_points` back to txmempool.h and updating lockpoints within the lambda (original approach in b5de6cc10bfbc8174936c3e8ae9418dedfb60000). Should I open a pull with that?

hebasto replied:
> Actually [bedf246](https://github.com/bitcoin/bitcoin/commit/bedf246f1e2497a3641093c6e8fa11fb34dddac4) is wrong.
> 
> `CheckSequenceLocks` modifies `lp`, so dropping it and using a fresh `lp` here from `GetLockPoints` is illegal.

#23897 aims to make it less fragile in the future.

MarcoFalke commented on src/txmempool.cpp:
```diff
@@ -642,7 +642,7 @@ void CTxMemPool::removeForReorg(CChainState& active_chainstate, int flags)
         const CTransaction& tx = it->GetTx();
         LockPoints lp = it->GetLockPoints();
-        bool validLP =  TestLockPointValidity(active_chainstate.m_chain, &lp);
+        const bool validLP = TestLockPointValidity(active_chainstate.m_chain, &lp);
```
nit: Could use `{}` instead of `=` to enable more compile time checks.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -350,8 +350,39 @@ void CChainState::MaybeUpdateMempoolForReorg(
+        AssertLockHeld(::cs_main);
+        const CTransaction& tx = it->GetTx();
+        LockPoints lp = it->GetLockPoints();
+        bool validLP = TestLockPointValidity(m_chain, &lp);
```
why add a `const` first and then remove it again?

MarcoFalke commented on src/validation.cpp:
```diff
@@ -350,8 +350,39 @@ void CChainState::MaybeUpdateMempoolForReorg(
+                const Coin &coin = CoinsTip().AccessCoin(txin.prevout);
+                assert(!coin.IsSpent());
+                unsigned int nMemPoolHeight = m_chain.Tip()->nHeight + 1;
+                if (coin.IsSpent() || (coin.IsCoinBase() && ((signed long)nMemPoolHeight) - coin.nHeight < COINBASE_MATURITY)) {
```
unrelated, but it seems a bit odd to have an integer and cast it three times for no reason. `int`->`unsigned`->`signed long`

glozow replied:
Done in #23649

maflcko reviewed (commented):
left comments

MarcoFalke commented:
This is not refactoring, but introducing a bug: https://github.com/bitcoin/bitcoin/pull/22677#discussion_r762040101
